File: ClickLight.Windows/ClickEvent.cs
namespace ClickLight.Windows;

/// <summary>
/// Types of mouse click events captured by the hook.
/// </summary>
public enum ClickKind
{
    LeftDown,
    LeftUp,
    RightDown,
    RightUp,
    Drag,
    Move,
    FileDrag
}

/// <summary>
/// Represents a single captured click event with location and timing.
/// </summary>
public sealed class ClickEvent
{
    public ClickKind Kind { get; }
    public double X { get; }
    public double Y { get; }
    public double Timestamp { get; }

    /// <summary>
    /// For Drag events: the point where the drag started (physical screen pixels).
    /// </summary>
    public double DragStartX { get; }
    public double DragStartY { get; }

    public ClickEvent(ClickKind kind, double x, double y, double timestamp, double dragStartX = 0, double dragStartY = 0)
    {
        Kind = kind;
        X = x;
        Y = y;
        Timestamp = timestamp;
        DragStartX = dragStartX;
        DragStartY = dragStartY;
    }

    public bool IsRelease => Kind == ClickKind.LeftUp || Kind == ClickKind.RightUp;
}

File: Sources/ClickLight/ClickEvent.swift
import AppKit

enum ClickKind: Sendable {
    case leftDown
    case leftUp
    case rightDown
    case rightUp
    case middleDown
    case middleUp
    case drag
    case move

    var isRelease: Bool {
        self == .leftUp || self == .rightUp || self == .middleUp
    }
}

struct ClickEvent: Sendable {
    let kind: ClickKind
    let location: CGPoint
    let timestamp: TimeInterval
}

struct KeyboardShortcutEvent: Sendable {
    let displayString: String
    let location: CGPoint
    let timestamp: TimeInterval
}

final class ClickEventBox: @unchecked Sendable {
    let event: ClickEvent

    init(_ event: ClickEvent) {
        self.event = event
    }
}

final class KeyboardShortcutEventBox: @unchecked Sendable {
    let event: KeyboardShortcutEvent

    init(_ event: KeyboardShortcutEvent) {
        self.event = event
    }
}
